pennylane/transforms/core/transform_program.py

```diff
@@ -92,15 +92,19 @@ def _jax_jac(classical_function, argnums, *args, **kwargs) -> qml.typing.TensorL
 
 
 # pylint: disable=unused-argument
-def _classical_preprocessing(qnode, program, tape_idx: int, *args, argnums=None, **kwargs):
+def _classical_preprocessing(qnode, program, tape_idx: int, *args, **kwargs):
     """Returns the trainable gate parameters for a given QNode input.
 
     While differentiating this again for each tape in the batch may be less efficient than desireable for large batches,
     it cleanly works with all interfaces.
     """
     tape = qml.workflow.construct_tape(qnode, level=0)(*args, **kwargs)
     tapes, _ = program((tape,))
-    return math.stack(tapes[tape_idx].get_parameters(trainable_only=True))
+    new_tape = tapes[tape_idx]
+    # Recalculate trainable parameters because the program is not obligated to maintain them
+    params = new_tape.get_parameters(trainable_only=False)
+    new_tape.trainable_params = qml.math.get_trainable_indices(params)
+    return math.stack(new_tape.get_parameters(trainable_only=True))
```

Comment:
Additional bug fix that ensures that trainability information is recomputed in case an intermediate transform did not preserve it.

Reply:
It does actually. Probably better to remove the specific fix for `validate_device_Wires`, because it makes the transform obey a contract we don't have and its effect is anyways overwritten by recomputing.